src/main/java/com/rocketden/main/service/RoomService.java

```diff
@@ -298,4 +299,36 @@ public RoomDto updateRoomSettings(String roomId, UpdateSettingsRequest request)
         socketService.sendSocketUpdate(roomDto);
         return roomDto;
     }
+
+    public RoomDto setSpectator(String roomId, SetSpectatorRequest request) {
+        Room room = repository.findRoomByRoomId(roomId);
+
+        // Return error if room could not be found
+        if (room == null) {
+            throw new ApiException(RoomError.NOT_FOUND);
+        }
+
+        //Return error if the initiator or receiver are null
+        if (request.getInitiator() == null || request.getReceiver() == null) {
+            throw new ApiException(UserError.NOT_FOUND);
+        }
+
+        //Return error if the initiator or receiver are not in the room
+        if (room.getUserByUserId(request.getReceiver().getUserId()) == null
+            || room.getUserByUserId(request.getInitiator().getUserId()) == null) {
+            throw new ApiException(RoomError.USER_NOT_FOUND);
+        }
+        // Return error if the initiator is not the host or the same as the receiver
+        User initiator = UserMapper.toEntity(request.getInitiator());
+        User receiver = UserMapper.toEntity(request.getReceiver());
+        if (!room.getHost().equals(initiator) && !initiator.equals(receiver)) {
+            throw new ApiException(RoomError.INVALID_PERMISSIONS);
+        }
+
+        User modifiedUser = room.getUserByUserId(receiver.getUserId());
+        modifiedUser.setSpectator(request.getSpectator());
+        repository.save(room);
```

Comment:
Check that `request.getSpectator() == null` above this code (and throw `BAD_SETTING` if so); otherwise, the spectator in the backend can equal `null` on a bad request, which we want to avoid.